src/storage/secondary/rowset/rowset_iterator.rs

```diff
@@ -126,26 +125,48 @@ impl RowSetIterator {
         // TODO: parallel fetch
         // TODO: align unmatched rows
 
-        for id in 0..filter_column.len() {
-            if filter_column[id] {
-                if let Some((row_id, array)) = self.column_iterators[id]
-                    .as_mut()
-                    .unwrap()
-                    .next_batch(Some(fetch_size), None)
-                    .await?
-                {
-                    if let Some(x) = common_chunk_range {
-                        if x != (row_id, array.len()) {
-                            panic!("unmatched rowid from column iterator");
+        let mut visibility_map = None;
+
+        for (id, column_ref) in self.column_refs.iter().enumerate() {
+            if let Some((_, filter_column)) = filter_context {
+                // Filter needed, so only filtered columns should be filled currently
+                if !filter_column[id] {
+                    arrays.push(None);
+                    continue;
+                }
+            }
+            // Normal procedure
+            match column_ref {
+                StorageColumnRef::RowHandler => arrays.push(None),
+                StorageColumnRef::Idx(_) => {
+                    if let Some((row_id, array)) = self.column_iterators[id]
+                        .as_mut()
+                        .unwrap()
+                        .next_batch(Some(fetch_size), visibility_map.as_ref())
+                        .await?
+                    {
+                        if let Some(x) = common_chunk_range {
+                            if x != (row_id, array.len()) {
+                                panic!("unmatched rowid from column iterator");
+                            }
+                        } else if !self.dvs.is_empty() {
+                            // Apply delete vector here to reduce the data
+                            // fetched next during scan
+                            let mut visi = BitVec::with_capacity(array.len());
+                            for _ in 0..array.len() {
+                                visi.push(true);
+                            }
+                            for dv in &self.dvs {
+                                dv.apply_to(&mut visi, row_id);
```

Comment:
We can produce the delete vector visibility map before scanning the filter column? And we may only need to apply delete vector once :)

Reply:
I think we are on the same track, the only problem is that I didn't know how to get the start row_id in this batch without scanning 1 column first, which made some codes confusing... Here the `dv` would only be applied once as well, I guess. Maybe we could add an interface `get_current_row_id` for the column_iterator, and call it on the first column to generate the visibility map before doin scan? And after that, adjust the following code logic according to your comment, which would make the logic more clear.

Reply:
> Maybe we could add an interface get_current_row_id for the column_iterator

We definitely need a `current_row_id`, but there are multiple places we can add it.

> and call it on the first column

We can also record the `current_row_id` in `RowSetIterator`. See:

https://github.com/risinglightdb/risinglight/blob/53a3388ec5866290d6d2d01065121e41c372c025/src/storage/secondary/rowset/rowset_iterator.rs#L37-L40

Currently, the `current_row_id` always starts at 0.

You may choose the best place to add the `current_row_id` variable.

Reply:
After a second thought, maybe adding it on `ColumnIterator` is a better way.

Reply:
Agree, since we already have https://github.com/risinglightdb/risinglight/blob/53a3388ec5/src/storage/secondary/column/concrete_column_iterator.rs#L46, the only work is to encapsulate a method in `ColumnIterator `

Reply:
It seems that we already have `get_current_row_id`, but only exposed for `test` 🤪